src/jdk.unsupported/share/classes/sun/misc/Unsafe.java

```diff
@@ -1255,12 +1731,207 @@ public void fullFence() {
+            return;
+        }
+
+        // warn && first usage, debug, or deny
+        beforeMemoryAccessSlow();
+    }
+
+    private static void beforeMemoryAccessSlow() {
+        assert MEMORY_ACCESS_OPTION != MemoryAccessOption.ALLOW;
+
+        // stack trace without the frames for the beforeMemoryAccess methods
+        List<StackWalker.StackFrame> stack = StackWalkerHolder.INSTANCE.walk(s ->
+                s.dropWhile(f -> (f.getDeclaringClass() == Unsafe.class)
+                                && f.getMethodName().startsWith("beforeMemoryAccess"))
+                    .limit(32)
+                    .toList()
+        );
+
+        // callerClass -> Unsafe.methodName
+        String methodName = stack.get(0).getMethodName();
+        Class<?> callerClass = stack.get(1).getDeclaringClass();
+
+        switch (MEMORY_ACCESS_OPTION) {
+            case WARN -> {
+                if (trySetMemoryAccessWarned()) {
+                    log(multiLineWarning(callerClass, methodName));
+                }
+            }
+            case DEBUG -> {
+                String warning = singleLineWarning(callerClass, methodName);
+                StringBuilder sb = new StringBuilder(warning);
+                stack.stream()
+                        .skip(1)
+                        .forEach(f ->
+                                sb.append(System.lineSeparator()).append("\tat " + f)
+                        );
+                log(sb.toString());
+            }
+            case DENY -> {
+                throw new UnsupportedOperationException(methodName);
+            }
+        }
+    }
+
+    /**
+     * Represents the options for the depreacted method-access methods.
+     */
+    private enum MemoryAccessOption {
+        /**
+         * Allow use of the memory-access methods with no warnings.
+         */
+        ALLOW,
+        /**
+         * Warning on the first use of a memory-access method.
+         */
+        WARN,
+        /**
+         * One-line warning and a stack trace on every use of a memory-access method.
+         */
+        DEBUG,
+        /**
+         * Deny use of the memory-access methods.
+         */
+        DENY;
+
+        private static MemoryAccessOption defaultValue() {
+            return ALLOW;
+        }
+
+        /**
+         * Return the value.
+         */
+        static MemoryAccessOption value() {
+            String value = VM.getSavedProperty("sun.misc.unsafe.memory.access");
+            if (value != null) {
+                return switch (value) {
+                    case "allow" -> MemoryAccessOption.ALLOW;
+                    case "warn"  -> MemoryAccessOption.WARN;
+                    case "debug" -> MemoryAccessOption.DEBUG;
+                    case "deny"  -> MemoryAccessOption.DENY;
+                    default -> {
+                        // emit warning
+                        log("--sun-misc-unsafe-memory-access ignored, value '" + value +
```

Comment:
The FFM code throws if an unknown value is passed. Here we log. Should we try to be more consistent?

Reply:
> The FFM code throws if an unknown value is passed. Here we log. Should we try to be more consistent?

I don't have a strong opinion on this. The value of --illegal-native-access is examined during startup so startup can fail if a bad value is specified. Here it is mid-flight when Unsafe initialises. The coin toss landed on printing a warning but it could equally fail. I jus didn't want code executing during startup to check configuration for something faraway in the jdk.unsupported module.